application/src/main/kotlin/no/kartverket/matrikkel/bygning/application/models/BygningEgenregistreringAggregering.kt

```diff
@@ -56,26 +57,57 @@ fun Bruksenhet.applyEgenregistrering(egenregistrering: Egenregistrering): Brukse
         vannforsyning = this.vannforsyning.aggregate(bruksenhetRegistrering.vannforsyningRegistrering) {
             Vannforsyning(
                 data = it.vannforsyning,
-                metadata = metadata.withKildemateriale(it.kildemateriale),
+                metadata = metadata
+                    .withKildemateriale(it.kildemateriale)
+                    .withGyldighetsperiode(
+                        Gyldighetsperiode.of(gyldighetsaar = it.gyldighetsaar, opphoersaar = it.opphoersaar),
+                    ),
             )
         },
         avlop = this.avlop.aggregate(bruksenhetRegistrering.avlopRegistrering) {
             Avlop(
                 data = it.avlop,
-                metadata = metadata.withKildemateriale(it.kildemateriale),
+                metadata = metadata
+                    .withKildemateriale(it.kildemateriale)
+                    .withGyldighetsperiode(
+                        Gyldighetsperiode.of(gyldighetsaar = it.gyldighetsaar, opphoersaar = it.opphoersaar),
+                    ),
             )
         },
+
+        // TODO Generalisering her?
         energikilder = this.energikilder.aggregate(bruksenhetRegistrering.energikildeRegistrering) {
-            Energikilde(
-                data = it.energikilder,
-                metadata = metadata.withKildemateriale(it.kildemateriale),
-            )
+            val currentEnergikilder = energikilder.egenregistrert ?: emptyList()
+
+            val (newEnergikilderToUpdate, newEnergikilderToAdd) = it.partition { registrering ->
+                registrering.energikilde in currentEnergikilder.map { current -> current.data }
+            }
+
+            currentEnergikilder
+                .map { energikilde ->
+                    newEnergikilderToUpdate.find { it.energikilde == energikilde.data }?.toEnergikilde(metadata)
+                        ?: energikilde
+                }
+                .plus(
+                    newEnergikilderToAdd.map { it.toEnergikilde(metadata) },
+                )
         },
-        oppvarminger = this.oppvarminger.aggregate(bruksenhetRegistrering.oppvarmingRegistrering) {
-            Oppvarming(
-                data = it.oppvarminger,
-                metadata = metadata.withKildemateriale(it.kildemateriale),
-            )
+
+        oppvarming = this.oppvarming.aggregate(bruksenhetRegistrering.oppvarmingRegistrering) {
+            val currentOppvarming = oppvarming.egenregistrert ?: emptyList()
+
+            val (newOppvarmingToUpdate, newOppvarmingToAdd) = it.partition { registrering ->
+                registrering.oppvarming in currentOppvarming.map { current -> current.data }
+            }
+
+            currentOppvarming
+                .map { oppvarming ->
+                    newOppvarmingToUpdate.find { it.oppvarming == oppvarming.data }?.toOppvarming(metadata)
+                        ?: oppvarming
+                }
```

Comment:
Kunne dette blitt forenklet slik?

```suggestion
            currentOppvarming
                .mapNotNull { oppvarming ->
                    newOppvarmingToUpdate.find { it.oppvarming == oppvarming.data }?.toOppvarming(metadata)
                }
```

Reply:
Mulig jeg kimser litt, men det vil vel ikke være det samme?

Nå sier vi at vi:
1. Tar dagens oppvarming
2. Sjekker om den samme koden har blitt sendt og eventuelt mapper
3. Hvis den ikke har blitt sendt inn, behold dagens oppvarming

Så om vi har `current == [Elektrisk(år=a), Sentralvarme(år=b)]`, og `newToUpdate == [Sentralvarme(år=c)]` så får vi til slutt `[Elektrisk(år=a), Sentralvarme(år=c)]`, mens `mapNotNull` ville returnert `[Sentralvarme(år=c)]`, right?

Reply:
Du har helt rett, jeg som blingsa!

Kunne man gjort koden noe mer lesbar med noe ala dette f.eks.?

```
currentOppvarming
    .map { oppvarming ->
        oppdaterMedNyeVerdierEllerBrukEksisterende(newOppvarmingToUpdate, oppvarming, metadata)
    }
    .plus(
        newOppvarmingToAdd.map { it.toOppvarming(metadata) },
    )
```